### PR TITLE
Chapter 13 Exercises.

### PR DESCRIPTION
**Pull request by gpoirier**

### CODE REVIEW

File: exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.specs2.mutable._
+import org.specs2.matcher.Scope
+
+class IOSpec extends Specification {
+
+  class IO3Fixture extends Scope {
+    import IO3._
+    import Console._
+
+    def countDown(from: Int): Free[Console, Unit] = {
```
Comment (MasseGuillaume) [inline exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala:12]: This is also a useful function for testing recursion complexity:
http://en.wikipedia.org/wiki/Ackermann_function#Table_of_values

File: exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      else
+        printLn("0.")
+    }
+  }
+
+  "runConsoleFunction0" >> {
+    "is not stack-safe" >> new IO3Fixture {
+      import IO3._
+
+      runConsoleFunction0(countDown(100000)).apply should throwAn[StackOverflowError]
```
Comment (MasseGuillaume) [inline exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala:27]: https://github.com/MasseGuillaume/TTFI/blob/a7d3c2314c27e020a78f6e68b7173fa61efab637/ttfi-core/src/test/scala/FixSpecs.scala#L17

File: exercises/src/main/scala/fpinscala/iomonad/IO.scala [hunk 3]
```diff
@@ -306,18 +308,41 @@ object IO3 {
                                f: A => Free[F, B]) extends Free[F, B]
 
   // Exercise 1: Implement the free monad
-  def freeMonad[F[_]]: Monad[({type f[a] = Free[F,a]})#f] = ???
+  def freeMonad[F[_]]: Monad[({type f[a] = Free[F,a]})#f] = new Monad[({type f[a] = iomonad.Free[F, a]})#f] {
```
Comment (MasseGuillaume) [inline exercises/src/main/scala/fpinscala/iomonad/IO.scala]: Take a look at
https://github.com/MasseGuillaume/reasonably-priced-monad
https://www.parleys.com/tutorial/composable-application-architecture-reasonably-priced-monads

Comment (MasseGuillaume) [inline exercises/src/main/scala/fpinscala/iomonad/IO.scala]: You should take a look at https://github.com/non/kind-projector it really saves you this boilerplate:

```
def freeMonad[F[_]]: Monad[({type f[a] = Free[F,a]})#f] = new Monad[({type f[a] = iomonad.Free[F, a]})#f]
```

vs

```
def freeMonad[F[_]]: Monad[Free[F,?]] = new Monad[Free[F,?]]
```

File: exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala (hunk 2 above)
Comment (gpoirier) [inline exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala:27]: I'm not quite sure what you're trying to say with this link?

Comment (MasseGuillaume) [inline exercises/src/test/scala/fpinscala/iomonad/IOSpec.scala:27]: two things: 
1) ackerman function usage example.
2) if you wanted to create a matcher like that you need to invert the order to make it work.

File: exercises/src/main/scala/fpinscala/iomonad/IO.scala (hunk 3 above)
Comment (gpoirier) [inline exercises/src/main/scala/fpinscala/iomonad/IO.scala]: I agree it's sounds interesting, and I'm sure I would use it on real code. But the type defined like that was already in the book's repo, it's autocompletion, I didn't write that.

**Pull request merged**